util/src/MetricsServer.cpp

```diff
@@ -0,0 +1,144 @@
+// Concord
+//
+// Copyright (c) 2019 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#include <string.h>
+#include <unistd.h>
+#include <iostream>
+
+#include "MetricsServer.hpp"
+
+// The MetricServer only handles a single type of request, that requests a JSON
+// metrics blob. The message format is just a single 8-bit 0 on the wire. We can
+// always change the protocol later if needed.
+const uint8_t kRequest = 0;
+
+// The MetricsServer only returns a single type of response. The response is
+// just an 8-bit 1 on the wire followed by a JSON string containing all metrics
+// for all components. Since we are using UDP, the entire message will always be
+// included, so no need to worry about framing. We can always change the
+// protocol if we decide to enhance the Metric server later on or move to a
+// different transport.
+const uint8_t kReply = 1;
+const uint8_t kError = 2;
+
+namespace concordMetrics {
+
+void Server::Start() {
+  if ((sock_ = socket(AF_INET, SOCK_DGRAM, 0)) < 0) {
+    LOG_FATAL(logger_, "Error creating UDP socket");
+    exit(1);
+  }
+
+  struct sockaddr_in servaddr;
+  memset(&servaddr, 0, sizeof(servaddr));
+  servaddr.sin_family = AF_INET;
+  servaddr.sin_addr.s_addr = INADDR_ANY;
+  servaddr.sin_port = htons(listenPort_);
+
+  if (bind(sock_, (const struct sockaddr*)&servaddr, sizeof(servaddr)) < 0) {
+    LOG_FATAL(logger_,
+              "Error binding UDP socket: IP=" << servaddr.sin_addr.s_addr
+                                              << ", Port=" << servaddr.sin_port
+                                              << ", errno=" << strerror(errno));
+    exit(1);
+  }
+
+  running_lock_.lock();
+  running_ = true;
+  running_lock_.unlock();
+
+  auto recvThread = std::thread(&Server::RecvLoop, &*this);
+  std::swap(thread_, recvThread);
+}
+
+void Server::Stop() {
```

Comment:
I would check the running_ here as well, although less critical since the return value of close() is not used

Reply:
The `thread_.join` call will fail if there is no running thread like so: `libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: thread::join failed: Invalid argument`

A bigger problem is calling join twice which may deadlock. But `running_` will already be set to false so this check won't help. In general this thread is supposed to run for the lifetime of the program and only be stopped on shutdown. It's relatively unlikely for a bug like this to pop up.